### PR TITLE
adds a basic cloudsmith plugin for backstage

### PR DESCRIPTION
**Pull request by punkle**
adds a basic cloudsmith plugin for backstage

This adds a homepage plugin, to show the stats for a cloudsmith repository.

image

#### :heavy_check_mark: Checklist

- [x] Added tests for new functionality and regression tests for bug fixes
- [x] Added changeset (run `yarn changeset` in the root)
- [x] Screenshots of before and after attached (for UI changes)
- [x] Added or updated documentation (if applicable)

### CODE REVIEW

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * A component to render the stats about a cloudsmith repo
+ *
+ * @public
+ */
+export const Content = ({ owner, repo }: CloudsmithStatsCardContentProps) => {
+  const fetchApi = useApi(fetchApiRef);
+  const discoveryApi = useApi(discoveryApiRef);
+  const cloudsmithApi = new CloudsmithClient({ fetchApi, discoveryApi });
+  const [repoStats, setRepoStats] = useState<RepoStats | undefined>();
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:50]:
```suggestion

```

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  const { loading, error } = useAsync(async () => {
+    setRepoStats(await cloudsmithApi.getRepoMetrics({ owner, repo }));
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:53]:
```suggestion
   return cloudsmithApi.getRepoMetrics({ owner, repo });
```

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * @public
+ */
+export const Content = ({ owner, repo }: CloudsmithStatsCardContentProps) => {
+  const fetchApi = useApi(fetchApiRef);
+  const discoveryApi = useApi(discoveryApiRef);
+  const cloudsmithApi = new CloudsmithClient({ fetchApi, discoveryApi });
+  const [repoStats, setRepoStats] = useState<RepoStats | undefined>();
+
+  const { loading, error } = useAsync(async () => {
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:52]:
```suggestion
  const { value: repoStats, loading, error } = useAsync(async () => {
```

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * A component to render the stats about a cloudsmith repo
+ *
+ * @public
+ */
+export const Content = ({ owner, repo }: CloudsmithStatsCardContentProps) => {
+  const fetchApi = useApi(fetchApiRef);
+  const discoveryApi = useApi(discoveryApiRef);
+  const cloudsmithApi = new CloudsmithClient({ fetchApi, discoveryApi });
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:49]: Cannot we do something like this ? That we ask for the cloudsmithApi with the cloudsmithApiRef ? 
```
  import { CloudsmithClient, RepoStats, cloudsmithApiRef } from '../../api';
  ...
```
```suggestion
  const cloudsmithApi = useApi(cloudsmithApiRef);
```

File: plugins/frontend/backstage-plugin-cloudsmith/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+Edit the `packages/app/src/components/home/Homepage.tsx` and add the imports
+
+Then add the following components:
+
+```typescript jsx
+<Grid item md={6}>
+  <CloudsmithStatsCard repo="repo" owner="owner" />
+</Grid>
+```
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/README.md:13]: Can you please be more descriptive in the README ? This is going to be the only documentation about this plugin. 
Can you add pls images of the plugin ? How will it look like if configured properly ? 
Can you please add a section about that they'll need to configure a proxy for it to work ? 
Can you please explain what is 'repo' and 'owner' ?
Can you please mention that they'll need to have homepage enabled (it is not a default feature of backstage still.) And link to the docs how to do it?

**Review by kissmikijr — CHANGES_REQUESTED** [review]
some tiny things

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx (hunk 4 above)
Comment (punkle) [reply plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:49]: It doesnt seem to work for home plugins, I am not sure why.

**Comment by punkle** [comment]
@kissmikijr done all those things. I dont seem to be able to get an API working in the component. I am not sure why.

File: plugins/frontend/backstage-plugin-cloudsmith/src/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import { createCardExtension } from '@backstage/plugin-home';
+
+import { CloudsmithClient, cloudsmithApiRef } from './api';
+
+export const cloudsmithPlugin = createPlugin({
+  id: 'cloudsmith',
+  apis: [
+    createApiFactory({
+      api: cloudsmithApiRef,
+      deps: { fetchApi: fetchApiRef, discoveryApi: discoveryApiRef },
```
Comment (Irma12) [inline plugins/frontend/backstage-plugin-cloudsmith/src/plugin.ts:34]: You might be missing
`      factory: ({ discoveryApi, fetchApi }) => new CloudsmithClient({ discoveryApi, fetchApi})
`

Comment (punkle) [reply plugins/frontend/backstage-plugin-cloudsmith/src/plugin.ts:34]: Isnt that on the next line?

Comment (Irma12) [reply plugins/frontend/backstage-plugin-cloudsmith/src/plugin.ts:34]: I didn't see it, sorry 😟

File: plugins/frontend/backstage-plugin-cloudsmith/src/api/CloudsmithApi.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { createApiRef } from '@backstage/core-plugin-api';
+
+export const cloudsmithApiRef = createApiRef<CloudsmithApi>({
```
Comment (Irma12) [inline plugins/frontend/backstage-plugin-cloudsmith/src/api/CloudsmithApi.ts:19]: shouldn't this be createApiRef<CloudsmithClient>

Comment (punkle) [reply plugins/frontend/backstage-plugin-cloudsmith/src/api/CloudsmithApi.ts:19]: I'm not sure what you mean exactly.

Comment (Irma12) [reply plugins/frontend/backstage-plugin-cloudsmith/src/api/CloudsmithApi.ts:19]: Missing the important part. 
I meant to write createApiRef<CloudsmithClient>

Comment (punkle) [reply plugins/frontend/backstage-plugin-cloudsmith/src/api/CloudsmithApi.ts:19]: I think it is correct already.

File: plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+  const {
+    value: repoStats,
+    loading,
+    error,
+  } = useAsync(async () => {
+    return await cloudsmithApi.getRepoMetrics({ owner, repo });
```
Comment (kissmikijr) [inline plugins/frontend/backstage-plugin-cloudsmith/src/components/CloudsmithStatsCard/Content.tsx:56]: I don't think we need to await this promise, as the function is an async function anyways.

**Review by kissmikijr — APPROVED** [review]

**Pull request merged**